File: epi/cpp/knapsack.cc
#include <unordered_map>
#include <vector>

#include "test_framework/generic_test.h"
#include "test_framework/serialization_traits.h"

#define MAX(a, b) (a > b ? a : b);

using std::pair;
using std::unordered_map;
using std::vector;

struct Item {
  int weight, value;
};

int OptimumSubjectToCapacity(const vector<Item>& items, int capacity) {
  int cap_to_maxval[items.size() + 1][capacity + 1];
  for (int j = 0; j <= capacity; j++) {
    cap_to_maxval[0][j] = 0;
  }

  for (int i = 1; i <= items.size(); i++) {
    int cweight = items[i - 1].weight, cval = items[i - 1].value;
    for (int j = 0; j <= capacity; j++) {
      int prev_val = cap_to_maxval[i - 1][j];
      int use_curr =
          (j >= cweight) ? cap_to_maxval[i - 1][j - cweight] + cval : 0;
      cap_to_maxval[i][j] = MAX(prev_val, use_curr);
    }
  }
  return cap_to_maxval[items.size()][capacity];
}

namespace test_framework {
template <>
struct SerializationTrait<Item> : UserSerTrait<Item, int, int> {};
}  // namespace test_framework

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"items", "capacity"};
  return GenericTestMain(args, "knapsack.cc", "knapsack.tsv",
                         &OptimumSubjectToCapacity, DefaultComparator{},
                         param_names);
}

File: epi/python/knapsack.py
import collections
import functools
from typing import List

from test_framework import generic_test
from test_framework.test_utils import enable_executor_hook

Item = collections.namedtuple('Item', ('weight', 'value'))


def optimum_subject_to_capacity(items: List[Item], capacity: int) -> int:
    best_val = [[0 for _ in range(capacity+1)] for _ in range(len(items)+1)]
    for i, item in enumerate(items):
        curr_i = i+1
        for curr_cap in range(capacity+1):
            best_val[curr_i][curr_cap] = best_val[curr_i - 1][curr_cap]
            if curr_cap >= item.weight:
                best_val[curr_i][curr_cap] = max(best_val[curr_i][curr_cap], best_val[curr_i-1][curr_cap - item.weight] + item.value)
    return best_val[-1][-1]


@enable_executor_hook
def optimum_subject_to_capacity_wrapper(executor, items, capacity):
    items = [Item(*i) for i in items]
    return executor.run(
        functools.partial(optimum_subject_to_capacity, items, capacity))


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('knapsack.py', 'knapsack.tsv',
                                       optimum_subject_to_capacity_wrapper))
